### PR TITLE
Upgrade raiden-contracts to 0.29.0

### PR DESCRIPTION
**Pull request by pirapira**
The changes are mostly
* adding another argument to TokenNetwork.closeChannel() function
* reading another parameter from TokenNetwork.ChannelSettled() event
* changing the calls to ServiceRegistry, whose implementation changed completely

### CODE REVIEW

File: raiden/network/proxies/token.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 from raiden.network.rpc.transactions import check_transaction_threw
 from raiden.utils import safe_gas_limit
 from raiden.utils.typing import Address, Balance, BlockSpecification, TokenAddress, TokenAmount
-from raiden_contracts.constants import CONTRACT_HUMAN_STANDARD_TOKEN
+from raiden_contracts.constants import CONTRACT_CUSTOM_TOKEN
```
Comment (pirapira) [inline raiden/network/proxies/token.py]: I really doubt I can do this here.

File: raiden/network/proxies/service_registry.py [hunk 2]
```diff
@@ -75,6 +86,27 @@ def get_service_url(
...
 
+    def current_price(self, block_identifier: BlockSpecification) -> int:
+        """Gets the currently required deposit amount."""
+
+        return self.proxy.contract.functions.currentPrice().call(block_identifier=block_identifier)
+
+    def token_address(self, block_identifier: BlockSpecification) -> AddressHex:
+        return self.proxy.contract.functions.token().call(block_identifier=block_identifier)
+
+    def deposit(self, limit_amount: int) -> None:
```
Comment (palango) [inline raiden/network/proxies/service_registry.py]: Do we need this here? Only the services should ever care, right?

Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: There is a test case in the client that tests `setURL()`.  Now a deposit is needed before `setURL()`.

**Comment by pirapira** [comment]
> reading another parameter from TokenNetwork.ChannelSettled() event

I didn't do this but the tests pass.  Deferred to https://github.com/raiden-network/raiden/issues/4472

File: raiden/network/proxies/token_network.py [hunk 3]
```diff
@@ -1425,7 +1425,8 @@ def close(
         balance_hash: BalanceHash,
         nonce: Nonce,
         additional_hash: AdditionalHash,
-        signature: Signature,
+        non_closing_signature: Signature,
```
Comment (LefterisJP) [inline raiden/network/proxies/token_network.py]: As discussed in the chat, we have a code freeze for the release. I would like to avoid any unnecessary changes for the release at the moment. So unless this change is needed for this release I would like to avoid it.

**important note**: I can't find anything wrong with this code in my review though.

But the proxies are used everywhere so it's not a small client change and bug can sip in at any point.

File: raiden/tests/integration/network/proxies/test_payment_channel.py [hunk 4]
```diff
@@ -101,28 +117,33 @@ def test_payment_channel_proxy_basics(
...
         channel_proxy_1.close(
             nonce=0,
             balance_hash=EMPTY_HASH,
             additional_hash=EMPTY_HASH,
-            signature=EMPTY_SIGNATURE,
+            non_closing_signature=EMPTY_SIGNATURE,
+            closing_signature=LocalSigner(private_keys[1]).sign(data=closing_data),
             block_identifier="latest",
         )
+        pytest.fail(msg)
```
Comment (LefterisJP) [inline raiden/tests/integration/network/proxies/test_payment_channel.py]: why remove the msg from `raises()`? Due to the deprecation I guess: https://docs.pytest.org/en/latest/deprecations.html#raises-message-deprecated ?

Not a bad change but please keep changes to a bare minimum when making a PR. A PR should do one thing only and anything else is a distraction. Both for the author but also for the reviewer/s

File: raiden/tests/integration/network/transport/test_matrix_transport.py [hunk 5]
```diff
@@ -318,14 +318,15 @@ def test_matrix_message_sync(matrix_transports):
         assert any(getattr(m, "message_identifier", -1) == i for m in received_messages)
 
 
-@pytest.mark.skipif(getattr(pytest, "config").getvalue("usepdb"), reason="test fails with pdb")
 @pytest.mark.parametrize("number_of_nodes", [2])
 @pytest.mark.parametrize("channels_per_node", [1])
 @pytest.mark.parametrize("number_of_tokens", [1])
 def test_matrix_tx_error_handling(  # pylint: disable=unused-argument
-    raiden_chain, token_addresses
+    raiden_chain, token_addresses, request
```
Comment (LefterisJP) [inline raiden/tests/integration/network/transport/test_matrix_transport.py]: same comment as above to keeping unrelated changes out of PRs

File: requirements/requirements-ci.txt [hunk 6]
```diff
@@ -123,27 +120,27 @@ pycryptodome==3.8.2
...
+pyrsistent==0.15.3
+pysaml2==4.8.0
 pysha3==1.0.2
 pytest-forked==1.0.2
 pytest-random==0.2
 pytest-select==0.1.2
-pytest-xdist==1.28.0
-pytest==4.6.3
+pytest-xdist==1.29.0
+pytest==5.0.1
```
Comment (LefterisJP) [inline requirements/requirements-ci.txt]: Please refrain from making such unrelated changes inside such a big PR. You are switching us to a new major pytest release. Why introduce this change in this PR?

File: requirements/requirements-ci.txt [hunk 7]
```diff
@@ -123,27 +120,27 @@ pycryptodome==3.8.2
...
 python-dateutil==2.8.0
 python-magic==0.4.15      # via s3cmd
-pytoml==0.1.20
+pytoml==0.1.21
 pytz==2019.1
 pyyaml==5.1.1
-raiden-contracts==0.26.0
-raiden-webui==0.8.0
+raiden-contracts==0.29.0
+raiden-webui==0.9.1
```
Comment (LefterisJP) [inline requirements/requirements-ci.txt]: Why upgrade the raiden-webui in this PR?

File: requirements/requirements-ci.txt [hunk 8]
```diff
@@ -48,38 +48,36 @@ eth-keys==0.2.4
...
 filelock==3.0.12
 flake8-bugbear==19.3.0
 flake8-tuple==0.4.0
-flake8==3.7.7
+flake8==3.7.8
 flaky==3.6.0
 flask-cors==3.0.8
 flask-restful==0.3.7
-flask==1.0.3
+flask==1.1.1
```
Comment (LefterisJP) [inline requirements/requirements-ci.txt]: why change flask version?

File: requirements/requirements-ci.txt [hunk 9]
```diff
@@ -123,27 +120,27 @@ pycryptodome==3.8.2
 pyflakes==2.1.1
 pygments==2.4.2
 pyhamcrest==1.9.0
-pyinstaller==3.4
+pyinstaller==3.5
```
Comment (LefterisJP) [inline requirements/requirements-ci.txt]: why change pyinstaller version in this PR?

File: requirements/requirements-dev.txt [hunk 10]
```diff
@@ -17,13 +17,13 @@ attrs==19.1.0             # via automat, black, flake8-bugbear, hypothesis, json
 automat==0.7.0            # via twisted
 autopep8==1.4.4
 babel==2.7.0
-bcrypt==3.1.6             # via matrix-synapse
+bcrypt==3.1.7             # via matrix-synapse
```
Comment (LefterisJP) [inline requirements/requirements-dev.txt]: Hmm there are way too many dependency changes. Why are you introducing them in this PR? Perhaps you guys changed something in the way dependencies are handled while I was gone?

File: raiden/network/proxies/token_network.py (hunk 3 above)
Comment (pirapira) [reply raiden/network/proxies/token_network.py]: @LefterisJP it's possible to remove this change.  `TokenNetwork` just needs to leave `closeChannel()` unchanged, and add a new `closeChannelByThirdParty()`.

File: raiden/tests/integration/network/proxies/test_payment_channel.py (hunk 4 above)
Comment (pirapira) [reply raiden/tests/integration/network/proxies/test_payment_channel.py]: I didn't know `requirements/deps upgrade raiden-contracts` was possible.  Instead, I ran `requirements/deps upgrade`.  I'll fix this in this PR.

File: raiden/network/proxies/token_network.py (hunk 3 above)
Comment (pirapira) [reply raiden/network/proxies/token_network.py]: @LefterisJP shall I do ^?

File: requirements/requirements-ci.txt (hunk 6 above)
Comment (pirapira) [reply requirements/requirements-ci.txt]: I didn't know `requirements/dep upgrade` can take a package name.  I'll remove unrelated changes in this PR.

File: requirements/requirements-ci.txt (hunk 7 above)
Comment (pirapira) [reply requirements/requirements-ci.txt]: I didn't know `requirements/dep upgrade` can take a package name.  Will fix.

File: requirements/requirements-ci.txt (hunk 8 above)
Comment (pirapira) [reply requirements/requirements-ci.txt]: I didn't know `requirements/dep upgrade` can take a package name.  Will fix.

File: requirements/requirements-ci.txt (hunk 9 above)
Comment (pirapira) [reply requirements/requirements-ci.txt]: I didn't know `requirements/dep upgrade` can take a package name.  Will fix.

File: requirements/requirements-dev.txt (hunk 10 above)
Comment (pirapira) [reply requirements/requirements-dev.txt]: I didn't know `requirements/dep upgrade` can take a package name.  Will fix.

File: raiden/network/proxies/token.py (hunk 1 above)
Comment (pirapira) [reply raiden/network/proxies/token.py]: During tests, I have to mint tokens for deposits in ServiceRegistry.

Shall I write a new proxy custom_token.py for tests?

File: raiden/network/proxies/service_registry.py [hunk 11]
```diff
@@ -75,6 +86,27 @@ def get_service_url(
             return None
         return result
 
+    def current_price(self, block_identifier: BlockSpecification) -> int:
+        """Gets the currently required deposit amount."""
+
```
[the remark this answers is not in the document]
Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: There is no justification for this.  I'll remove the line from the commit.

File: raiden/network/proxies/service_registry.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: * I'll add `block_identifier` here.
* I'll replace `int` with `TokenAmount` here, and in the return type of `current_price()`.

File: raiden/network/proxies/token.py (hunk 1 above)
Comment (pirapira) [reply raiden/network/proxies/token.py]: @palango is it important to have `setURL` function tested in this client repo?  Otherwise, I remove the test, and avoid introducing the test custom token here.

Comment (karlb) [reply raiden/network/proxies/token.py]: It will not be called by the client at all, so it makes sense to test it only in the contracts repo.

Comment (pirapira) [reply raiden/network/proxies/token.py]: Alright!  That simplifies a lot.

File: raiden/network/proxies/token_network.py (hunk 3 above)
Comment (pirapira) [reply raiden/network/proxies/token_network.py]: @Dominik1999 said I should keep this.

File: raiden/network/proxies/service_registry.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: Maybe I can remove these functions altogether.

File: raiden/tests/integration/network/transport/test_matrix_transport.py (hunk 5 above)
Comment (pirapira) [reply raiden/tests/integration/network/transport/test_matrix_transport.py]: This was also due to the deprecated `pytest.config`.  I'll revert the change.

File: raiden/network/proxies/token.py (hunk 1 above)
Comment (pirapira) [reply raiden/network/proxies/token.py]: However, in order to keep `test_service_registry_random_pfs` and `test_configure_pfs`, I need a `ServiceRegistry` with some services registered on it.

File: raiden/network/proxies/service_registry.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: No, some useful tests use this function.

File: raiden/network/proxies/token.py (hunk 1 above)
Comment (pirapira) [reply raiden/network/proxies/token.py]: So I'm keeping `CUSTOM_TOKEN` here.

Comment (LefterisJP) [reply raiden/network/proxies/token.py]: As discussed in the chat it should not affect the production code since the ABI is a superset of ERC-20.

File: raiden/network/pathfinding.py [hunk 12]
```diff
@@ -167,11 +167,25 @@ def get_random_pfs(
...
     If there are no PFS in the given registry, it returns (None, None).
     """
-    count = service_registry.service_count(block_identifier=block_identifier)
-    if count == 0:
+    count = service_registry.ever_made_deposits_len(block_identifier=block_identifier)
+    addresses = (
+        service_registry.ever_made_deposits(block_identifier=block_identifier, index=idx)
+        for idx in range(0, count)
+    )
+    address_list = list(
```
Comment (hackaugusto) [inline raiden/network/pathfinding.py]: This can be greatly simplified with a list comprehension:
```python
address_list = [
  addr
  for addr in addresses
  if service_registry.has_valid_registration(address=addr, block_identifier=block_identifier)
]
```

Comment (pirapira) [reply raiden/network/pathfinding.py]: OK.  I take that idea.

Comment (hackaugusto) [reply raiden/network/pathfinding.py]: Actually, the above can be even simpler with:

```python
addresses = []
for index in range(count):
    address = service_registry.ever_made_deposits(
        block_identifier=block_identifier,
        index=index,
    )
    is_valid = service_registry.has_valid_registration(
        address=address,
        block_identifier=block_identifier,
    )
    if is_valid:
        addresses.append(address)
```

If we could have ordered addresses it would be even simpler.

Comment (hackaugusto) [reply raiden/network/pathfinding.py]: Okay, third proposal, we can switch it around and reduce the number of RPC calls:

```python
number_of_addresses = service_registry.ever_made_deposits_len(
    block_identifier=block_identifier,
)
addresses_to_try = list(range(number_of_addresses))
random.shuffle(addresses_to_try)

while addresses_to_try:
    index = addresses_to_try.pop()
    address = service_registry.ever_made_deposits(
        block_identifier=block_identifier,
        index=index,
    )
    is_valid = service_registry.has_valid_registration(
        address=address,
        block_identifier=block_identifier,
    )
    if is_valid:
        return address
return None
```

File: raiden/network/proxies/service_registry.py [hunk 13]
```diff
@@ -45,19 +46,29 @@ def __init__(
         self.client = jsonrpc_client
         self.node_address = self.client.address
 
-    def service_count(self, block_identifier: BlockSpecification) -> int:
-        """Get the number of registered services"""
-        result = self.proxy.contract.functions.serviceCount().call(
+    def ever_made_deposits(
+        self, block_identifier: BlockSpecification, index: int
+    ) -> Optional[AddressHex]:
```
Comment (hackaugusto) [inline raiden/network/proxies/service_registry.py]: IMO this should be bytes, to respect the sandwhich encoding.

File: raiden/network/proxies/service_registry.py [hunk 14]
```diff
@@ -45,19 +46,29 @@ def __init__(
...
             block_identifier=block_identifier
         )
         return result
 
-    def get_service_address(
-        self, block_identifier: BlockSpecification, index: int
-    ) -> Optional[AddressHex]:
-        """Gets the address of a service by index. If index is out of range return None"""
+    def has_valid_registration(
+        self, block_identifier: BlockSpecification, address: AddressHex
```
Comment (hackaugusto) [inline raiden/network/proxies/service_registry.py]: here too the input should be binary.

File: raiden/network/proxies/service_registry.py (hunk 13 above)
Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: OK.  Will follow this.

File: raiden/network/proxies/service_registry.py (hunk 14 above)
Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: Here too I change the types.

File: raiden/network/pathfinding.py (hunk 12 above)
Comment (pirapira) [reply raiden/network/pathfinding.py]: I'll do this with `indices_to_try`.

File: raiden/network/proxies/token_network.py [hunk 15]
```diff
@@ -1444,20 +1445,38 @@ def close(
...
-                token_network_address=self.address,
-                channel_identifier=channel_identifier,
-            )
+        canonical_identifier = CanonicalIdentifier(
+            chain_identifier=self.proxy.contract.functions.chain_id().call(),
+            token_network_address=self.address,
+            channel_identifier=channel_identifier,
+        )
+
+        our_signed_data = pack_balance_proof_update(
```
Comment (pirapira) [inline raiden/network/proxies/token_network.py]: The name `pack_balance_proof_update` sounds obsolete because it is now used not only for balanceProofUpdate calls, but also for closeChannels() call. Is  `pack_signed_balance_proof` a better name?

File: raiden/network/pathfinding.py (hunk 12 above)
Comment (pirapira) [reply raiden/network/pathfinding.py]: I followed this somehow.

File: raiden/network/proxies/token_network.py (hunk 15 above)
Comment (LefterisJP) [reply raiden/network/proxies/token_network.py]: Hmm you have a point. I think the name you propose sounds better. What do you think @hackaugusto ?

Comment (hackaugusto) [reply raiden/network/proxies/token_network.py]: good with me :)

File: raiden/network/proxies/service_registry.py [hunk 16]
```diff
@@ -48,7 +48,7 @@ def __init__(
 
     def ever_made_deposits(
         self, block_identifier: BlockSpecification, index: int
-    ) -> Optional[AddressHex]:
+    ) -> Optional[Address]:
```
Comment (LefterisJP) [inline raiden/network/proxies/service_registry.py]: Ah but this won't work alone like this. You also need to transform the hexadecimal address returned by the call to bytes from below. You need to use `to_canonical_address()`

Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: I added `assert isinstance(ret, T_Address)` and trying locally to confirm failure.

File: raiden/network/proxies/token_network.py (hunk 15 above)
Comment (pirapira) [reply raiden/network/proxies/token_network.py]: Then I do replace-string.

File: raiden/network/proxies/service_registry.py (hunk 16 above)
Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: Nice
```
raiden/network/pathfinding.py:192: error: Argument "service_hex_address" to "get_service_url" of "ServiceRegistry" has incompatible type "Address"; expected "AddressHex"
```

Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: Also
```
raiden/network/proxies/service_registry.py:65: error: Incompatible return value type (got "bytes", expected "Optional[Address]")
```

Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: Done the necessary changes.

File: raiden/network/proxies/service_registry.py [hunk 17]
```diff
@@ -45,19 +53,36 @@ def __init__(
...
+        """Get one of the addresses that have ever made a deposit."""
+        try:
+            ret = Address(
+                to_canonical_address(
+                    self.proxy.contract.functions.ever_made_deposits(index).call(
+                        block_identifier=block_identifier
+                    )
+                )
+            )
+            assert isinstance(ret, T_Address)
```
Comment (LefterisJP) [inline raiden/network/proxies/service_registry.py]: Why is the assertion here?

File: raiden/network/proxies/service_registry.py [hunk 18]
```diff
@@ -45,19 +53,36 @@ def __init__(
...
-    def get_service_address(
-        self, block_identifier: BlockSpecification, index: int
-    ) -> Optional[AddressHex]:
-        """Gets the address of a service by index. If index is out of range return None"""
+    def has_valid_registration(
+        self, block_identifier: BlockSpecification, address: Address
+    ) -> Optional[bool]:
         try:
-            result = self.proxy.contract.functions.service_addresses(index).call(
+            assert isinstance(address, T_Address)
```
Comment (LefterisJP) [inline raiden/network/proxies/service_registry.py]: same here. I guess we wanted to not have runtime type assertions on every function, that's why we switched to mypy.

File: raiden/network/proxies/service_registry.py [hunk 19]
```diff
@@ -75,6 +100,26 @@ def get_service_url(
             return None
         return result
 
+    def current_price(self, block_identifier: BlockSpecification) -> TokenAmount:
+        """Gets the currently required deposit amount."""
+        return self.proxy.contract.functions.currentPrice().call(block_identifier=block_identifier)
+
+    def token_address(self, block_identifier: BlockSpecification) -> AddressHex:
```
Comment (LefterisJP) [inline raiden/network/proxies/service_registry.py]: Can you also return the bytes address here? We need to be consistent.

File: raiden/network/proxies/service_registry.py (hunk 17 above)
Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: Once I didn't see `mypy` catching this (I had an older version of `mypy` or something).  But now I've seen `mypy` working.  So I can remove this.

File: raiden/network/proxies/service_registry.py (hunk 18 above)
Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: The same here (locally my `mypy` was not working).  Will remove.

File: raiden/network/proxies/service_registry.py (hunk 19 above)
Comment (pirapira) [reply raiden/network/proxies/service_registry.py]: OK!

File: raiden/tests/utils/smartcontracts.py [hunk 20]
```diff
@@ -84,26 +87,61 @@ def deploy_service_registry_and_set_urls(
         service_registry_address=service_registry_address,
         contract_manager=contract_manager,
     )
+    token_address = decode_hex(c1_service_proxy.token_address(block_identifier="latest"))
```
Comment (LefterisJP) [inline raiden/tests/utils/smartcontracts.py]: @pirapira this line (and maybe others) need adjusting for the tests to pass now that token_address is also bytes

Comment (pirapira) [reply raiden/tests/utils/smartcontracts.py]: Interesting.  `decode_hex` is maybe not type annotated.

**Comment by Dominik1999** [comment]
that PR looks like a lot of work :) Thanks for pushing that Yoichi

File: raiden/tests/utils/smartcontracts.py (hunk 20 above)
Comment (karlb) [reply raiden/tests/utils/smartcontracts.py]: Not in the eth_utils version we're using. Unfortunately, upgrading to a version with type annotations is not trivial.
